### ISSUE TITLE
Create another branch for maintenance

### ISSUE THREAD

**Issue by tuhdo** [syl20bnr/spacemacs#1064]
We should create another branches for merging bug fixes and independent features, so we can release Spacemacs more often without being blocked by major changes in develop branch.

**Comment by trishume** [comment]
Or we could use the standard Git and git-flow practice of feature branches. For example this release would have had a bunch of PR merges and small things on develop which could have been released incrementally and then two major feature branches `holy-mode` and `auto-completion-overhaul` where all the work to get those working well and then merge them into `develop` when they are done.

**Comment by syl20bnr** [comment]
Thank you @tuhdo for the proposition, we highly need it.

My postmortem about the current release (which is not released yet ... well :-)) is that the develop branch turned into a mix of uncontrollable feature branches. We have the issue of all big scope feature branches where we never know when we can merge it back and when it will be ready. This is a **fine** state for the develop branch to me, the mistake was to not have a _release branch_ `0.100.1` to back port the fixes for the bugs that made people to switch on develop. 

I won't make a `0.100.1` branch though because it is too late, but there will be a `0.101.0` where will be free to cherry-pick vital fixes to allow the users to continue to use a stable release and be not annoyed about the noise of a develop branch.

My proposal is to keep it simple with the following branches:
- master: current release
- release `x.x.x`: maintenance branch
- develop: things can go wrong and this is OK.

So we end up with 3 release levels: Stable, Maintenance and Development. I see  good opportunity to mimic the distribution channels of Chrome here and to have an integration into spacemacs to switch between the channels easily, decoupling the git stuff behind it.

**Comment by trishume** [comment]
This would work fine but I think it would be easier and fit better with the existing system to use feature branches instead. That way things will work in a familiar way and the update system and the default git clone to install and update will continue to work.

I think develop should be "bleeding edge" but not broken, the equivalent of an alpha channel not a nightly. That's what people like us contributors want, we want to test out the latest things and shake out minor problems but preferably still be able to use it as a text editor. When develop is unstable but not broken like in the past it works great and we can do some testing and release every couple days. It has worked quite well generally.

What happened recently is that develop got some major new things that even @syl20bnr with his extraordinary productivity couldn't get completely working in one evening. In the past @syl20bnr has approximated the effect of him using feature branches by just being amazing and committing all the changes and fixes before anyone could checkout the intervening broken commits.

Basically to avoid this in the future the only change that needs to happen is that @syl20bnr should use a feature branch when doing any big changes. We basically already follow this model 90%, most of us use feature branches and PRs that go through iteration and code review and only get merged when they seem to be done. But @syl20bnr works differently, and that's fine except in this case.

We could all use the same system and solve the problem that way, like most Github projects do, or if @syl20bnr likes to work differently that could work too, as long as there is some other solution. I just like this one since it fits best with the existing system and other Github projects while still solving the problem.

**Comment by syl20bnr** [comment]
I could use a branch for my work but then again nobody will use it and it just delay the regression detection because I cannot test all the languages.

We don't have to make a worst portrait of the situation than it really is because it leads to over-complicated branch design for little benefits.

There is no such feature that is vital for one user to have right now. Waiting for a couple of days to be in master should be the de facto thing to do and I won't put in place some facilities for people that cannot wait for a couple of days. Develop is the develop branch and it can be broken, we had the proof this week and it is fine to me.

If it's true that there is no vital feature it is also true that there is a lot of vital fixes. What is laking now is just a maintenance branch to be able to give them to the users in a timely fashion like we did up to now.

Nobody should use the develop branch and think that it won't break for his or her use cases.

**Comment by trishume** [comment]
Ok fair enough. I'm still not entirely sure what you mean by maintenance branch and how that would work. When a user clicks the update button in Spacemacs or clones the git repo what will they get? What will show on the Spacemacs git repo home page?

Instead of making it a separate branch and having a non-linear upgrade flow maybe we should accomplish the same thing in the [git-flow style](http://nvie.com/posts/a-successful-git-branching-model/) by just having one `hotfixes` branch that is merged into master as a new release and then incorporated in develop.

That way we don't have a million branches and the upgrade and `git clone` always gets you the latest in a linear stream of stable releases.

**Comment by syl20bnr** [comment]
I want to try a flow where all the merges are one-way only. I'm not sure yet if it will be successful or even doable or if we can find a cool way to use them for the users.

Basically here is the relations between the branches:
- develop -> master
- develop -> maintenance
- contributors branches -> develop

That's all. No bi-directional integration.

The long story:

Master is the receptacle for a new ready to release code base, when we are ready to throw the `x.x.x` version we merge develop into master. When we do this we create a new maintenance (release is a synonym) branch off master. Typically users should use the last maintenance branch, not the master. The maintenance branch dies when it lags behind master after say... 5 new releases (i.e. commits in master). In the maintenance branch we are free to cherry-pick or back port stuff of the develop branch, the master branch won't see them until the next deliverable code base. Minor versions can be fork of the current maintenant branch.

So all in all we move from a tag base release model to a branch name based release model.

**Comment by syl20bnr** [comment]
We the above model we can ditch on type of branch...

Master would be develop :-) And we roll only with release branches. Getting a new version of spacemacs is just a matter of checking out a release branch.

**Comment by trishume** [comment]
I'm not sure I like that, with this system:
- The main Github page no longer shows the latest stable info
- `git clone` no longer installs the latest stable thing
- We have to redo all the update logic
- We lose Github showing nice changelogs
- Each bug fix must be backported to N different other branches.
- It's an unconventional way to do a Github project so it might confuse new users/contributors

With some effort a couple of these might be regained, but what do we gain over the git-flow model with a hotfix branch?

**Comment by syl20bnr** [comment]
- The main branch of a repo can be changed.
- If we keep master then it will a stable release, then at spacemacs startup we could automatically upgrade
- not a problem to redo, we cannot always get things right the first time
- nope change log can be made on any branch
- nope they are backported only to the current release
- maybe but it's 3 lines of text to explain it, done :-)

**Comment by syl20bnr** [comment]
For the last question: simplicity.

**Comment by trishume** [comment]
Ok thanks. Automatically updating to the latest on the branch to fix bugs without people having to come to Gitter is a convincing benefit. Another I just thought of is now the Github PR thing will default to the correct branch for PRs!

Also I take back Github showing the latest stable thing, this is something I don't like currently. People looking at code are usually contributors who want to see the latest develop. I certainly often find myself hitting the branch switch button on Github.

And as long as we can figure out some good install instructions that get the latest stable release it should be good.

**Comment by syl20bnr** [comment]
This model will work only if it is transparent for the user. We should be able to do it, the user will ever see only release numbers.

There are some tweaks to be done to make it even more fancy but I wrote the general idea of it... oh well I can write one: the minor version could be the number of commits on the maintenance branch since its creation. So we have automatic tracking of the specific version of a user without bothering with minor versions. We should just care about the `b` in the `a.b.c` version number.

I would like to put develop as the default branch but the documentation won't reflect the current stable which can be confusing :-( I have the same issue as you with constantly switch to develop.

**Comment by trishume** [comment]
@syl20bnr I imagine you plan on scripting the Github API a bunch then? I certainly wouldn't want to constantly be changing the default branch.

**Comment by syl20bnr** [comment]
From a release manager perspective yes, from a user perspective, onyl plain git commands (see the current update notifications, just plain git commands).

**Comment by trishume** [comment]
Also you could solve the docs problem by rendering the markdown somewhere else, maybe Github pages. A custom page with a sidebar stylesheet would also solve the problem of the first N screens of documentation being table of contents and having to scroll back up to reach it again.

**Comment by syl20bnr** [comment]
Great ideas, I definitely have to do this page, would be a great way to celebrate the 1k stars.

**Comment by tuhdo** [comment]
Originally I thought that we create a maintenance branch and when a contributor pushes a change to Spacemacs, he will add a title prefix such as `[maintenance]` or `[develop]` to know which is bug fix and which is new feature. If something is labeled as `[maintenance]`, it is merged into both `maintenance` and `develop` branches. If something is labeled as `[develop]`, then it is merged into `develop` branch.

I think this can work well with the model @syl20bnr described above.

**Comment by syl20bnr** [comment]
Indeed, very good proposition. The `CONTRIBUTE.md` file needs some work to be less technical (I can handle the technical aspect of cherry-picking etc.... the contributors does not care about that) and make things important like PR titles more obvious.
